File: src/onomedit/core/pipeline.py
"""重命名主流程编排：收集 → 编辑器 → 读回 → 应用规则 → 执行 → 日志。"""

from __future__ import annotations

import difflib
import os
from dataclasses import dataclass, field
from pathlib import Path

from onomedit.core import (
    collection,
    editor,
    envvars,
    rules,
    tempfile_mgr,
)
from onomedit.core import (
    config as config_mod,
)
from onomedit.core.logger import RenameLogger, parse_line
from onomedit.core.pathitem import PathItem, rename_ensure_parent
from onomedit.i18n import tr
from onomedit.utils import safename


class PipelineError(RuntimeError):
    """主流程中止级错误（无文件、未配置编辑器、行数校验失败等）。"""


class DuplicateTargetError(PipelineError):
    """目标重名：编辑结果中多个文件指向同一目标路径，中止执行（不执行任何重命名）。"""

    def __init__(self, conflicts: list[tuple[str, list[str]]]):
        self.conflicts = conflicts
        # 每组冲突：目标单独一行、每个源文件各占一行，避免长路径挤在同一行
        lines = [tr("检测到目标重名，已中止（未执行任何重命名）:")]
        for new, olds in conflicts:
            lines.append(tr("  目标: {target}", target=new))
            for old in olds:
                lines.append(f"    <- {old}")
        super().__init__("\n".join(lines))

    @property
    def summary(self) -> str:
        """状态栏用的一句话摘要。"""
        groups = len(self.conflicts)
        involved = sum(len(olds) for _, olds in self.conflicts)
        return tr(
            "检测到目标重名，已中止（未执行任何重命名）: {groups} 组目标、涉及 {files} 个文件",
            groups=groups,
            files=involved,
        )


def find_duplicate_targets(pairs: list[tuple[str, str]]) -> list[tuple[str, list[str]]]:
    """找出目标重名组：不同源文件最终指向同一目标路径。

    - Windows 文件系统大小写不敏感：``same.txt`` 与 ``SAME.TXT`` 视为同一目标。
    - 无变化项（``old == new``，保持原名）不算"重命名到重名目标"，不参与判定。

    返回 ``[(目标路径, [源路径, ...]), ...]``，按目标首次出现顺序，仅含重复组。
    """
    groups: dict[str, list[str]] = {}
    first: dict[str, str] = {}  # 规范化键 → 首次出现的原始目标写法（用于显示）
    for old, new in pairs:
        if old == new:
            continue
        key = os.path.normcase(new)
        first.setdefault(key, new)
        groups.setdefault(key, []).append(old)
    return [(first[key], groups[key]) for key in groups if len(groups[key]) > 1]


@dataclass
class RenameResult:
    """批量重命名结果：成功 / 失败 / 无变化（跳过）。"""

    success: list[tuple[str, str]] = field(default_factory=list)
    failed: list[tuple[str, str, str]] = field(
        default_factory=list
    )  # (old, new, error)
    skipped: list[str] = field(default_factory=list)  # 目标与源相同，无变化

    @property
    def total(self) -> int:
        return len(self.success) + len(self.failed) + len(self.skipped)

    @property
    def changed(self) -> int:
        return len(self.success) + len(self.failed)


@dataclass
class PreviewRow:
    """列表窗口/预览用：原文件名、新文件名、差异、距离。"""

    old: str
    new: str
    diff: str
    distance: int


@dataclass
class PipelineOutcome:
    result: RenameResult = field(default_factory=RenameResult)
    pairs: list[tuple[str, str]] = field(default_factory=list)
    preview: list[PreviewRow] | None = None
    dry_run: bool = False
    temp_path: str = ""


def levenshtein(a: str, b: str) -> int:
    """编辑距离（单行内存版）。"""
    if len(a) < len(b):
        a, b = b, a
    if not b:
        return len(a)
    prev = list(range(len(b) + 1))
    for i, ca in enumerate(a, 1):
        cur = [i]
        for j, cb in enumerate(b, 1):
            cur.append(min(prev[j] + 1, cur[j - 1] + 1, prev[j - 1] + (ca != cb)))
        prev = cur
    return prev[-1]


def diff_text(a: str, b: str) -> str:
    """单行字符级差异标注：删除 [-x-] 插入 [+y+]。"""
    sm = difflib.SequenceMatcher(None, a, b)
    parts: list[str] = []
    for tag, i1, i2, j1, j2 in sm.get_opcodes():
        if tag == "equal":
            parts.append(a[i1:i2])
        elif tag == "delete":
            parts.append(f"[-{a[i1:i2]}-]")
        elif tag == "insert":
            parts.append(f"[+{b[j1:j2]}+]")
        else:  # replace
            parts.append(f"[-{a[i1:i2]}-][+{b[j1:j2]}+]")
    return "".join(parts)


class Renamer:
    """批量重命名执行器：目标重复预检、真冲突序号化、链/环两阶段解环、单文件失败不中断。"""

    def __init__(self, log: RenameLogger | None = None):
        self.log = log
        self.result = RenameResult()
        self._pending: dict[str, str] = {}
        self._removed: set[str] = set()
        self._moved: dict[str, tuple[str, str]] = {}
        self._tmp_seq = 0

    def run(self, pairs: list[tuple[str, str]]) -> RenameResult:
        # 目标重名预检：发现重名必须中止（不执行任何重命名），
        # 避免"多个目标名相同，最终只改其中一个"。
        conflicts = find_duplicate_targets(pairs)
        if conflicts:
            raise DuplicateTargetError(conflicts)

        # 第一阶段：移走冲突目标 + 完成无冲突改名
        self._pending = {old: new for old, new in pairs}
        for old, new in pairs:
            if old in self._removed:
                continue  # 内容已被链处理移走，其落位在第二阶段
            self._do(old, new)

        # 第二阶段：被移走的内容落位到最终目标
        for tmp, (old, target) in list(self._moved.items()):
            try:
                if os.path.exists(target):
                    target = str(safename.unique_path(Path(target)))
                rename_ensure_parent(tmp, target)
                self.result.success.append((old, target))
                self._log_record(old, target)
            except OSError as e:
                self.result.failed.append((old, target, str(e)))
                self._log_error(f"{old} -> {target}: {e}")
        return self.result

    def _do(self, old: str, new: str) -> None:
        if old == new:
            self._pending.pop(old, None)
            self.result.skipped.append(old)
            return
        try:
            if os.path.exists(new) and new in self._pending:
                # 链/环：new 是本批次尚未执行的源 → 先把内容移出到临时名
                tmp = self._temp_name(new)
                rename_ensure_parent(new, tmp)
                self._removed.add(new)
                self._moved[tmp] = (new, self._pending.pop(new))
            if os.path.exists(new):
                # 真冲突：目标仍被占用且不属于本批次 → 安全序号化
                new = str(safename.unique_path(Path(new)))
            rename_ensure_parent(old, new)
            self._pending.pop(old, None)
            self.result.success.append((old, new))
            self._log_record(old, new)
        except OSError as e:
            self._pending.pop(old, None)
            self.result.failed.append((old, new, str(e)))
            self._log_error(f"{old} -> {new}: {e}")

    def _temp_name(self, target: str) -> str:
        parent = os.path.dirname(target)
        stem, ext = os.path.splitext(os.path.basename(target))
        while True:
            self._tmp_seq += 1
            candidate = os.path.join(
                parent, f".__onomedit_tmp_{os.getpid()}_{self._tmp_seq}_{stem}{ext}"
            )
            if not os.path.lexists(candidate):
                return candidate

    def _log_record(self, old: str, new: str) -> None:
        if self.log:
            self.log.record(old, new)

    def _log_error(self, message: str) -> None:
        if self.log:
            self.log.record_error(message)


def restore(
    logger: RenameLogger,
    *,
    all_history: bool = False,
    partial_lines: list[str] | None = None,
) -> RenameResult:
    """恢复流程：反向执行（新 → 旧），倒序避免改名链依赖冲突。"""
    if partial_lines is not None:
        pairs = [parse_line(line) for line in partial_lines]
    elif all_history:
        pairs = logger.read_history()
    else:
        pairs = logger.read_last()
    if not pairs:
        return RenameResult()
    reverse = [(new, old) for old, new in reversed(pairs)]
    renamer = Renamer(log=logger)
    return renamer.run(reverse)


class RenamePipeline:
    """主流程编排。所有阶段可被调用方注入（测试/多实例隔离）。"""

    def __init__(
        self,
        cfg=None,
        *,
        temp_dir=None,
        on_status=None,
        clipboard_text: str | None = None,
    ):
        self.cfg = cfg or config_mod.load_config()
        self.temp_dir = temp_dir or (self.cfg.temp_dir or None)
        self.on_status = on_status or (lambda msg: None)
        self.clipboard_text = clipboard_text

    def prepare(
        self, raw_paths: list[str] | None = None
    ) -> tuple[list[PathItem], list[str], str]:
        """流程前半：收集 → 展开 → 过滤 → 写临时文件 → 等待编辑 → 读回。

        返回 (items, new_full 列表, 临时文件路径)。行数校验失败抛 LineCountError。
        """
        paths = collection.collect_paths(
            raw_paths, use_clipboard=self.clipboard_text is None
        )
        if not paths:
            raise PipelineError(tr("没有可处理的文件（路径不存在或剪贴板为空）"))
        items = collection.build_items(paths)
        if self.cfg.expand_subdirs:
            items = collection.expand_subdirs(items, self.cfg.subdirs_depth)
        items = collection.apply_excludes(items, self.cfg.exclude)
        # 去重：多 glob 模式重叠、显式传同一路径、嵌套目录展开等均会产生重复项，
        # 重复项会导致同一文件出现在临时文件多行，重命名时相互冲突。
        items = collection.dedupe_items(items)
        items = collection.sort_items(
            items, self.cfg.sort_by, reverse=self.cfg.sort_reverse
        )
        if not items:
            raise PipelineError(tr("应用排除规则后没有可处理的文件"))

        temp_path, _ = tempfile_mgr.write_items(
            items, self.cfg.path_type, temp_dir=self.temp_dir
        )
        sig = tempfile_mgr.signature(temp_path)

        if self.cfg.open_editor:
            editor_cmd = self.cfg.editor
            if not editor_cmd.strip():
                raise PipelineError(
                    tr(
                        "未配置编辑器，请先运行: onomedit config set-editor <命令>\n"
                        "（或使用 --no-editor 跳过编辑器）"
                    )
                )
            self.on_status(
                tr(
                    "已写入临时文件: {path}\n请在编辑器中修改后保存并退出…",
                    path=temp_path,
                )
            )
            editor.launch_and_wait(
                editor_cmd,
                temp_path,
                sig,
                multi_tab=self.cfg.multi_tab,
                timeout=self.cfg.editor_timeout,
                on_status=self.on_status,
            )

        lines = tempfile_mgr.read_lines(
            temp_path, len(items)
        )  # 行数校验，不一致抛 LineCountError
        new_fulls = [
            item.with_field(self.cfg.path_type, line)
            for item, line in zip(items, lines)
        ]
        return items, new_fulls, temp_path

    def plan(
        self, items: list[PathItem], new_fulls: list[str]
    ) -> list[tuple[str, str]]:
        """应用规则（顺序固定）+ 环境变量展开 + 安全命名，生成 (old, new) 计划对。

        规则顺序：替换/转换/插入（按配置列表顺序）→ 环境变量展开 → 安全命名。
        """
        pairs: list[tuple[str, str]] = []
        env = envvars.EnvVars()
        for item, new_full in zip(items, new_fulls):
            full = new_full
            if self.cfg.apply_rules:
                if self.cfg.enable_auto_rules:
                    for rule in self.cfg.auto_rules:
                        pi = PathItem(full, is_dir=item.is_dir)
                        value = pi.get_field(rule.scope)
                        value = rules.apply_rule(value, rule)
                        full = pi.with_field(rule.scope, value)
                if self.cfg.enable_envvars:
                    # 环境变量作用于 name 段（目录段保持原样）
                    pi = PathItem(full, is_dir=item.is_dir)
                    ctx = envvars.EnvContext(
                        file=item.full, clip_text=self.clipboard_text
                    )
                    new_name = env.expand(pi.name, context=ctx)
                    full = os.path.join(pi.directory, new_name)
            if self.cfg.safety.sanitize:
                pi = PathItem(full, is_dir=item.is_dir)
                full = os.path.join(pi.directory, safename.sanitize_name(pi.name))
            pairs.append((item.full, full))
        return pairs

    def run_editor_mode(
        self,
        raw_paths: list[str] | None = None,
        *,
        dry_run: bool = False,
    ) -> PipelineOutcome:
        """编辑器模式全流程（流程 1）。dry_run 仅预览（差异/距离）不执行。"""
        items, new_fulls, temp_path = self.prepare(raw_paths)
        try:
            pairs = self.plan(items, new_fulls)
        finally:
            self._cleanup_temp(temp_path)

        if dry_run:
            rows = (
                preview_rows(items, pairs, self.cfg)
                if self.cfg.preview.diff or self.cfg.preview.distance
                else None
            )
            return PipelineOutcome(
                pairs=pairs, preview=rows, dry_run=True, temp_path=temp_path
            )

        log = RenameLogger(config_mod.log_dir())
        log.begin_session()
        renamer = Renamer(log=log)
        result = renamer.run(pairs)
        return PipelineOutcome(result=result, pairs=pairs, temp_path=temp_path)

    @staticmethod
    def _cleanup_temp(temp_path: str) -> None:
        try:
            os.unlink(temp_path)
        except OSError:
            pass


def preview_rows(
    items: list[PathItem], pairs: list[tuple[str, str]], cfg
) -> list[PreviewRow]:
    """生成列表窗口/预览所需行（差异 / 距离按配置开关）。"""
    rows: list[PreviewRow] = []
    for item, (old, new) in zip(items, pairs):
        rows.append(
            PreviewRow(
                old=old,
                new=new,
                diff=diff_text(old, new) if cfg.preview.diff else "",
                distance=levenshtein(old, new) if cfg.preview.distance else 0,
            )
        )
    return rows

File: crates/onomedit-core/src/pipeline.rs
use std::collections::{HashMap, HashSet};
use std::fs;
use std::io;
use std::path::{Path, PathBuf};

use thiserror::Error;

use crate::collection::{self, path_key};
use crate::config::Config;
use crate::diff::{diff_text, levenshtein};
use crate::edit_file::{EditFile, EditFileError};
use crate::journal::{RenameLogger, parse_line};
use crate::path::PathItem;
use crate::rules;
use crate::safe_name::{sanitize_name, unique_path};
use crate::template::{EnvContext, EnvVars};

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct RenamePair {
    pub old: PathBuf,
    pub requested_new: PathBuf,
}

impl RenamePair {
    pub fn new(old: impl Into<PathBuf>, new: impl Into<PathBuf>) -> Self {
        Self {
            old: old.into(),
            requested_new: new.into(),
        }
    }
}

#[derive(Debug, Clone, Default, PartialEq, Eq)]
pub struct RenameResult {
    pub success: Vec<(PathBuf, PathBuf)>,
    pub failed: Vec<(PathBuf, PathBuf, String)>,
    pub skipped: Vec<PathBuf>,
}

impl RenameResult {
    pub fn total(&self) -> usize {
        self.success.len() + self.failed.len() + self.skipped.len()
    }
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct DuplicateTargetError {
    pub conflicts: Vec<(PathBuf, Vec<PathBuf>)>,
}

impl DuplicateTargetError {
    pub fn summary(&self) -> String {
        let involved: usize = self
            .conflicts
            .iter()
            .map(|(_, sources)| sources.len())
            .sum();
        format!(
            "检测到目标重名，已中止（未执行任何重命名）: {} 组目标、涉及 {involved} 个文件",
            self.conflicts.len()
        )
    }
}

impl std::fmt::Display for DuplicateTargetError {
    fn fmt(&self, formatter: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        writeln!(formatter, "检测到目标重名，已中止（未执行任何重命名）:")?;
        for (index, (target, sources)) in self.conflicts.iter().enumerate() {
            writeln!(formatter, "  目标: {}", target.display())?;
            for (source_index, source) in sources.iter().enumerate() {
                let final_line =
                    index + 1 == self.conflicts.len() && source_index + 1 == sources.len();
                if final_line {
                    write!(formatter, "    <- {}", source.display())?;
                } else {
                    writeln!(formatter, "    <- {}", source.display())?;
                }
            }
        }
        Ok(())
    }
}

impl std::error::Error for DuplicateTargetError {}

#[derive(Debug, Error)]
pub enum PipelineError {
    #[error("没有可处理的文件（路径不存在或剪贴板为空）")]
    NoFiles,
    #[error("应用排除规则后没有可处理的文件")]
    NoFilesAfterExclude,
    #[error(transparent)]
    EditFile(#[from] EditFileError),
    #[error(transparent)]
    DuplicateTarget(#[from] DuplicateTargetError),
    #[error(transparent)]
    Io(#[from] io::Error),
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct PreviewRow {
    pub old: PathBuf,
    pub new: PathBuf,
    pub diff: String,
    pub distance: usize,
}

#[derive(Debug, Clone, Default)]
pub struct PipelineOutcome {
    pub result: RenameResult,
    pub pairs: Vec<RenamePair>,
    pub preview: Option<Vec<PreviewRow>>,
    pub dry_run: bool,
    pub temp_path: PathBuf,
}

pub struct PlanSession {
    pub items: Vec<PathItem>,
    edit_file: EditFile,
}

impl PlanSession {
    pub fn edit_path(&self) -> &Path {
        self.edit_file.path()
    }

    pub fn signature(&self) -> Result<crate::edit_file::Signature, EditFileError> {
        self.edit_file.signature()
    }
}

pub struct RenamePipeline {
    pub config: Config,
    pub clipboard_text: Option<String>,
}

impl RenamePipeline {
    pub fn new(config: Config) -> Self {
        Self {
            config,
            clipboard_text: None,
        }
    }

    pub fn with_clipboard_text(mut self, clipboard_text: Option<String>) -> Self {
        self.clipboard_text = clipboard_text;
        self
    }

    pub fn prepare(&self, raw_paths: &[String]) -> Result<PlanSession, PipelineError> {
        let paths = collection::collect_paths(raw_paths);
        if paths.is_empty() {
            return Err(PipelineError::NoFiles);
        }
        let mut items: Vec<PathItem> = paths.into_iter().map(PathItem::new).collect();
        if self.config.expand_subdirs {
            items = collection::expand_subdirs(items, self.config.subdirs_depth);
        }
        items = collection::apply_excludes(items, &self.config.exclude);
        items = collection::dedupe_items(items);
        items = collection::sort_items(items, &self.config.sort_by, self.config.sort_reverse);
        if items.is_empty() {
            return Err(PipelineError::NoFilesAfterExclude);
        }
        let temp_dir = (!self.config.temp_dir.is_empty()).then(|| Path::new(&self.config.temp_dir));
        let edit_file = EditFile::create(&items, self.config.path_type, temp_dir)?;
        Ok(PlanSession { items, edit_file })
    }

    pub fn finish_plan(&self, session: &PlanSession) -> Result<Vec<RenamePair>, PipelineError> {
        let lines = session.edit_file.read_lines(session.items.len())?;
        let edited: Vec<PathBuf> = session
            .items
            .iter()
            .zip(lines)
            .map(|(item, line)| item.with_field(self.config.path_type, &line))
            .collect();
        Ok(self.plan(&session.items, &edited))
    }

    pub fn plan(&self, items: &[PathItem], edited: &[PathBuf]) -> Vec<RenamePair> {
        let mut environment = EnvVars::default();
        items
            .iter()
            .zip(edited)
            .map(|(item, edited)| {
                let mut full = edited.clone();
                if self.config.apply_rules {
                    if self.config.enable_auto_rules {
                        for rule in &self.config.auto_rules {
                            let current = PathItem::new(&full);
                            let value = rules::apply(&current.field(rule.scope), rule);
                            full = current.with_field(rule.scope, &value);
                        }
                    }
                    if self.config.enable_envvars {
                        let current = PathItem::new(&full);
                        let context = EnvContext {
                            file: item.full().to_string_lossy().into_owned(),
                            clip_text: self.clipboard_text.clone(),
                        };
                        let name = environment.expand(&current.name(), Some(&context));
                        full = current.directory().join(name);
                    }
                }
                if self.config.safety.sanitize {
                    let current = PathItem::new(&full);
                    full = current
                        .directory()
                        .join(sanitize_name(&current.name(), "_"));
                }
                RenamePair::new(item.full(), full)
            })
            .collect()
    }

    pub fn run_direct(
        &self,
        raw_paths: &[String],
        dry_run: bool,
        logger: Option<RenameLogger>,
    ) -> Result<PipelineOutcome, PipelineError> {
        let session = self.prepare(raw_paths)?;
        let temp_path = session.edit_path().to_owned();
        let pairs = self.finish_plan(&session)?;
        if dry_run {
            let preview = (self.config.preview.diff || self.config.preview.distance)
                .then(|| preview_rows(&pairs, &self.config));
            return Ok(PipelineOutcome {
                pairs,
                preview,
                dry_run: true,
                temp_path,
                ..PipelineOutcome::default()
            });
        }
        if let Some(log) = &logger {
            log.begin_session();
        }
        let result = Renamer::new(logger).run(&pairs)?;
        Ok(PipelineOutcome {
            result,
            pairs,
            temp_path,
            ..PipelineOutcome::default()
        })
    }
}

pub fn preview_rows(pairs: &[RenamePair], config: &Config) -> Vec<PreviewRow> {
    pairs
        .iter()
        .map(|pair| {
            let old = pair.old.to_string_lossy();
            let new = pair.requested_new.to_string_lossy();
            PreviewRow {
                old: pair.old.clone(),
                new: pair.requested_new.clone(),
                diff: if config.preview.diff {
                    diff_text(&old, &new)
                } else {
                    String::new()
                },
                distance: if config.preview.distance {
                    levenshtein(&old, &new)
                } else {
                    0
                },
            }
        })
        .collect()
}

pub fn find_duplicate_targets(pairs: &[RenamePair]) -> Vec<(PathBuf, Vec<PathBuf>)> {
    let mut groups: Vec<(PathBuf, Vec<PathBuf>)> = Vec::new();
    let mut indexes = HashMap::new();
    for pair in pairs {
        if pair.old == pair.requested_new {
            continue;
        }
        let key = path_key(&pair.requested_new);
        let index = *indexes.entry(key).or_insert_with(|| {
            groups.push((pair.requested_new.clone(), Vec::new()));
            groups.len() - 1
        });
        groups[index].1.push(pair.old.clone());
    }
    groups.retain(|(_, sources)| sources.len() > 1);
    groups
}

pub struct Renamer {
    logger: Option<RenameLogger>,
    temp_sequence: usize,
}

impl Renamer {
    pub fn new(logger: Option<RenameLogger>) -> Self {
        Self {
            logger,
            temp_sequence: 0,
        }
    }

    pub fn run(mut self, pairs: &[RenamePair]) -> Result<RenameResult, DuplicateTargetError> {
        let conflicts = find_duplicate_targets(pairs);
        if !conflicts.is_empty() {
            return Err(DuplicateTargetError { conflicts });
        }

        let mut result = RenameResult::default();
        let mut pending: HashMap<String, RenamePair> = pairs
            .iter()
            .cloned()
            .map(|pair| (path_key(&pair.old), pair))
            .collect();
        let mut removed = HashSet::new();
        let mut moved: Vec<(PathBuf, PathBuf, PathBuf)> = Vec::new();

        for pair in pairs {
            let old_key = path_key(&pair.old);
            if removed.contains(&old_key) {
                continue;
            }
            if pair.old == pair.requested_new {
                pending.remove(&old_key);
                result.skipped.push(pair.old.clone());
                continue;
            }
            let mut actual_new = pair.requested_new.clone();
            let operation = (|| -> io::Result<()> {
                let new_key = path_key(&actual_new);
                if actual_new.exists() && pending.contains_key(&new_key) {
                    let displaced = pending.remove(&new_key).expect("checked pending key");
                    let temporary = self.temp_name(&actual_new);
                    rename_ensure_parent(&actual_new, &temporary)?;
                    removed.insert(new_key);
                    moved.push((temporary, displaced.old, displaced.requested_new));
                }
                if actual_new.exists() {
                    actual_new = unique_path(&actual_new);
                }
                rename_ensure_parent(&pair.old, &actual_new)
            })();
            pending.remove(&old_key);
            match operation {
                Ok(()) => {
                    result.success.push((pair.old.clone(), actual_new.clone()));
                    self.record(&pair.old, &actual_new);
                }
                Err(error) => {
                    result
                        .failed
                        .push((pair.old.clone(), actual_new.clone(), error.to_string()));
                    self.record_error(&format!(
                        "{} -> {}: {error}",
                        pair.old.display(),
                        actual_new.display()
                    ));
                }
            }
        }

        for (temporary, old, mut target) in moved {
            if target.exists() {
                target = unique_path(&target);
            }
            let operation = rename_ensure_parent(&temporary, &target);
            match operation {
                Ok(()) => {
                    result.success.push((old.clone(), target.clone()));
                    self.record(&old, &target);
                }
                Err(error) => {
                    result
                        .failed
                        .push((old.clone(), target.clone(), error.to_string()));
                    self.record_error(&format!(
                        "{} -> {}: {error}",
                        old.display(),
                        target.display()
                    ));
                }
            }
        }
        Ok(result)
    }

    fn temp_name(&mut self, target: &Path) -> PathBuf {
        let parent = target.parent().unwrap_or_else(|| Path::new(""));
        let stem = target
            .file_stem()
            .map(|v| v.to_string_lossy())
            .unwrap_or_default();
        let extension = target
            .extension()
            .map(|v| format!(".{}", v.to_string_lossy()))
            .unwrap_or_default();
        loop {
            self.temp_sequence += 1;
            let candidate = parent.join(format!(
                ".__onomedit_tmp_{}_{}_{}{}",
                std::process::id(),
                self.temp_sequence,
                stem,
                extension
            ));
            if !candidate.exists() {
                return candidate;
            }
        }
    }

    fn record(&self, old: &Path, new: &Path) {
        if let Some(logger) = &self.logger {
            logger.record(old, new);
        }
    }

    fn record_error(&self, message: &str) {
        if let Some(logger) = &self.logger {
            logger.record_error(message);
        }
    }
}

pub fn rename_ensure_parent(old: &Path, new: &Path) -> io::Result<()> {
    match fs::rename(old, new) {
        Ok(()) => Ok(()),
        Err(error) if error.kind() == io::ErrorKind::NotFound && old.exists() => {
            let Some(parent) = new.parent() else {
                return Err(error);
            };
            if parent.as_os_str().is_empty() {
                return Err(error);
            }
            fs::create_dir_all(parent)?;
            fs::rename(old, new)
        }
        Err(error) => Err(error),
    }
}

pub fn restore(
    logger: &RenameLogger,
    all_history: bool,
    partial_lines: Option<&[String]>,
) -> Result<RenameResult, PipelineError> {
    let pairs: Vec<(PathBuf, PathBuf)> = if let Some(lines) = partial_lines {
        lines
            .iter()
            .map(|line| parse_line(line))
            .collect::<Result<_, _>>()?
    } else if all_history {
        logger.read_history()
    } else {
        logger.read_last()
    };
    let reverse: Vec<RenamePair> = pairs
        .into_iter()
        .rev()
        .map(|(old, new)| RenamePair::new(new, old))
        .collect();
    Ok(Renamer::new(Some(logger.clone())).run(&reverse)?)
}

#[cfg(test)]
mod tests {
    use super::*;

    fn pair(old: &Path, new: &Path) -> RenamePair {
        RenamePair::new(old, new)
    }

    #[test]
    fn duplicate_target_aborts_before_changes() {
        let dir = tempfile::tempdir().unwrap();
        let a = dir.path().join("a.txt");
        let b = dir.path().join("b.txt");
        let same = dir.path().join("same.txt");
        fs::write(&a, "a").unwrap();
        fs::write(&b, "b").unwrap();
        let error = Renamer::new(None)
            .run(&[pair(&a, &same), pair(&b, &same)])
            .unwrap_err();
        assert_eq!(
            error.summary(),
            "检测到目标重名，已中止（未执行任何重命名）: 1 组目标、涉及 2 个文件"
        );
        assert!(a.exists() && b.exists() && !same.exists());
    }

    #[test]
    fn resolves_swap_cycle_without_temp_names_in_log() {
        let dir = tempfile::tempdir().unwrap();
        let a = dir.path().join("a.txt");
        let b = dir.path().join("b.txt");
        fs::write(&a, "AAA").unwrap();
        fs::write(&b, "BBB").unwrap();
        let logger = RenameLogger::new(dir.path().join("log"));
        logger.begin_session();
        let result = Renamer::new(Some(logger.clone()))
            .run(&[pair(&a, &b), pair(&b, &a)])
            .unwrap();
        assert_eq!(result.success.len(), 2);
        assert_eq!(fs::read_to_string(&a).unwrap(), "BBB");
        assert_eq!(fs::read_to_string(&b).unwrap(), "AAA");
        assert!(
            !fs::read_to_string(logger.last_path)
                .unwrap()
                .contains("__onomedit_tmp_")
        );
    }

    #[test]
    fn numbers_real_conflicts_and_creates_parent() {
        let dir = tempfile::tempdir().unwrap();
        let a = dir.path().join("a.txt");
        let b = dir.path().join("b.txt");
        fs::write(&a, "a").unwrap();
        fs::write(&b, "existing").unwrap();
        let result = Renamer::new(None).run(&[pair(&a, &b)]).unwrap();
        assert_eq!(result.success[0].1.file_name().unwrap(), "b (1).txt");
        let source = dir.path().join("source.txt");
        let target = dir.path().join("new/nested/target.txt");
        fs::write(&source, "x").unwrap();
        Renamer::new(None).run(&[pair(&source, &target)]).unwrap();
        assert!(target.exists());
    }
}
